### PR TITLE
fs-uae - added integration into emulationstation. Supporting ZIPs containing multiple disks

### PR DESCRIPTION
**Pull request by verybadsoldier**
So, me again ;)

Another try on integrating fs-uae into ES. Now taking the ZIP functions into account (so no multi-disk-discovery needed anymore).

Some explanations on some points (does not make them necessarily good decisions ofc):

-  I thought about adding Amiga to platforms.cfg and to store the supported extensions there. But the extensions are depending on the emulator. So fs-uae supports IPF, ADF etc. (and now ZIP) while uae4all seems to be ADF only. I don't know if platforms.cfg is supposed to handle this so I added the extensions as a variable to the emulator script. Actually I wanted to have it (FSUAE_EXTS) readonly but when I declare it readonly it is always empty later in the configure function. I could not find out why.

- I changed the config directory symbolic link from `~/.config/fs-uae` to `~/Documents/FS-UAE/Configuration` since this is the directory that fs-uae creates on first run to store and read the config from.

- I added a hopefully sensible default config for fs-uae to have it run in fullscreen by default. Also some other parameters are in there so they can be changed easily by the user.

Looking forward to your input!

### CODE REVIEW

File: scriptmodules/emulators/fs-uae.sh [hunk 1]
```diff
@@ -9,9 +9,11 @@
 # at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
 #
 
+FSUAE_EXTS=(.adf .ipf .dms .adz .zip)
```
Comment (joolswills) [inline scriptmodules/emulators/fs-uae.sh]: no globals here please.

File: scriptmodules/emulators/fs-uae.sh [hunk 2]
```diff
@@ -9,9 +9,11 @@
...
+
 rp_module_id="fs-uae"
 rp_module_desc="Amiga emulator - FS-UAE integrates the most accurate Amiga emulation code available from WinUAE"
-rp_module_help="ROM Extension: .adf"
+rp_module_help="ROM Extension: ${FSUAE_EXTS[*]}\n\nCopy your Amiga games to $romdir/amiga\n\nCopy a required BIOS file (e.g. kick13.rom) to $biosdir"
```
Comment (joolswills) [inline scriptmodules/emulators/fs-uae.sh]: I think I would prefer just putting the extensions in platforms.cfg - the emulators themselves document what they support and we have this in other cases where some emulators don't support all formats.

File: scriptmodules/emulators/fs-uae.sh [hunk 3]
```diff
@@ -47,16 +49,18 @@ function remove_fs-uae() {
 function configure_fs-uae() {
     mkRomDir "amiga"
 
-    mkUserDir "$home/.config"
+    # copy configuring start script
+    mkdir "$md_inst/bin/"
```
Comment (joolswills) [inline scriptmodules/emulators/fs-uae.sh]: minor, but no need for /

File: scriptmodules/emulators/fs-uae.sh [hunk 4]
```diff
@@ -47,16 +49,18 @@ function remove_fs-uae() {
...
+    cp "$scriptdir/scriptmodules/$md_type/$md_id/fs-uae.sh" "$md_inst/bin/"
```
Comment (joolswills) [inline scriptmodules/emulators/fs-uae.sh]: use `$md_data`

File: scriptmodules/emulators/fs-uae.sh [hunk 5]
```diff
@@ -47,16 +49,18 @@ function remove_fs-uae() {
...
+    moveConfigDir "$home/Documents/FS-UAE/Configurations" "$md_conf_root/amiga"
 
-    cat > "$romdir/amiga/+Start FS-UAE.sh" << _EOF_
-#!/bin/bash
-fs-uae-launcher
-_EOF_
-    chmod a+x "$romdir/amiga/+Start FS-UAE.sh"
-    chown $user:$user "$romdir/amiga/+Start FS-UAE.sh"
+    # copy default config file
+    cp -v "$scriptdir/scriptmodules/$md_type/$md_id/Default.fs-uae" "$md_conf_root/amiga"
```
Comment (joolswills) [inline scriptmodules/emulators/fs-uae.sh]: tiny config - can use iniSet - also make sure we don't overwrite existing configs on upgrade. (configure should not overwrite existing configs).

File: scriptmodules/emulators/fs-uae/fs-uae.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+exts="$1"
+rom="$2"
+
+rootdir="/opt/retropie"
+configdir="$rootdir/configs"
+config="$configdir/amiga/Default.fs-uae"
+
+user="$SUDO_USER"
+[[ -z "$user" ]] && user="$(id -un)"
```
Comment (joolswills) [inline scriptmodules/emulators/fs-uae/fs-uae.sh]: no need for this - the script is not run under sudo.

File: scriptmodules/emulators/fs-uae/fs-uae.sh [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+home="$(eval echo ~$user)"
```
Comment (joolswills) [inline scriptmodules/emulators/fs-uae/fs-uae.sh]: can just use `$HOME` in this case

**Comment by joolswills** [comment]
>I changed the config directory symbolic link from ~/.config/fs-uae to ~/Documents/FS-UAE/Configuration since this is the directory that fs-uae creates on first run to store and read the config from.

are you saying that `~/.config/fs-uae` is not used or created by fs-uae ?

**Comment by joolswills** [comment]
thanks - made a few comments above.

**Comment by verybadsoldier** [comment]
Ok, thanks for your hints, next iteration is here :)

Regarding config path:
> are you saying that ~/.config/fs-uae is not used or created by fs-uae ?

fs-uae reads from `Documents/FS-UAE/Configuration` but also from `~/.config/fs-uae`. The latter is the "Hidden Default Configuration File" as documented here: https://fs-uae.net/configuration-files
As I understand it, the main purpose of the hidden config is to use the parameter `base_dir` to redirect the whole fs-uae path to something custom.
So I think using `Documents` is the more straightforward way but I can also change it to `.config` if you prefer it.

I am still not really happy with the handling of the extensions. I have added them to platforms.cfg but also entered them in `rp_module_help`. Also the start script needs them at runtime. So I pass the extensions as parameter to the start script (read from platforms.cfg), because I think the start script cannot access platforms.cfg (and because platforms.cfg is not specific for the current emulator).
(In general I think it is a bit dirty for emulator scripts to use the extensions information from platforms.cfg because they are not specific for the emulator and also might change later when new emulators are added.)
I can live with it but if you have ideas how to improve it I would be happy to implement them.

Also when invoking fs-uae I removed the parameter which passes a specific configfile so I rely on fs-uae finding the default config (Default.fs-uae) on its own, which should be fine (hopefully). It simplified some lines of code in the start script.

**Comment by joolswills** [comment]
I'm happy with the extensions being in platforms.cfg.

Instead of passing by argument, you could include them directly in the launch script - by adding them in the same way we add extensions in the asplashscreen init.d script as it is installed. or put the script in as a heredoc as with vice, and put the extensions in the script there.

**Comment by joolswills** [comment]
Anyway, it's also fine as it is now, so I will merge - I may tweak a few things later.

**Comment by verybadsoldier** [comment]
Ok, great, again thanks!

**Pull request merged**